torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -0,0 +1,44 @@
+from typing import Any
+
+import torch.nn as nn
+
+from torch.distributed._composable import contract
+from torch.distributed._composable_state import _insert_module_state
+
+
+@contract()
+def fully_shard(
+    module: nn.Module,
+):
+    if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
+        raise ValueError(
+            f"fully_shard does not support containers that do not implement forward: {module}"
+        )
+    state = fully_shard.state(module)
+    _insert_module_state(module, state)
+    # Place FSDP leftmost for highest priority in the method resolution order
+    cls = module.__class__
+    dct = {"__deepcopy__": unimplemented_deepcopy}
+    new_cls = type(f"FSDP{cls.__name__}", (FSDP, cls), dct)
+    module.__class__ = new_cls
+    return module
+
+
+def unimplemented_deepcopy(*args: Any, **kwargs: Any) -> None:
+    raise AssertionError(
+        "FSDP does not support deepcopy. Please use state dict for serialization."
+    )
+
+
+class FSDP:
```

Comment:
is FSDP a convenience for people who are used to current FSDP to migrate? do we want this?  (why not just tell them use fully_shard)

Reply:
This class is used for us to add and override methods (see the PR description for explanation). It is distinct from `fully_shard` the frotend.

Reply:
Without a class like `FSDP`, we would have to monkey patch to add or override methods. Now, we can just define them on `FSDP` and since `FSDP` is earliest in the method resolution order, it will override any existing method definition.

Reply:
nvm i see that its used mainly as a target for isinstance checks.

mby worth documenting that it's not intended for manual use but is the class that your fully_shard(module)  takes on when sharded

Reply:
oh, your explanation helped even more. didn't see it when i typed the above. thanks!